feat(migration): add a script for automatically performing renaming in developers' files

BeksOmega opened the pull request.

BeksOmega committed:
feat: add basic setup for migration plugin

BeksOmega committed:
fix: update package name and README

BeksOmega committed:
fix: typos in readme

BeksOmega committed:
fix: remove bad test files

BeksOmega committed:
feat: add basic fetching

BeksOmega committed:
fix: Add inline docs to rename script

BeksOmega committed:
feat: add test infrastructure and reorganize rename script for testing

BeksOmega committed:
feat: add basic list of renaming tests

BeksOmega committed:
feat:add tests for renaming script

BeksOmega commented on plugins/migration/test/rename.mocha.js:
```diff
@@ -0,0 +1,527 @@
+import {doRenamings} from '../bin/rename.js';
+
+
+suite('Rename', function() {
```
I think this should be a pretty good start to a test suite. If you can think of any other cases that need to be covered I'm happy to add them.

cpcallen replied:
So far so good.  We've already thought of a few more and I'm sure will think of more yet, but a solid start!

BeksOmega commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,77 @@
+ * @param {!Array<string>} strings The strings to apply the renamings in.
+ * @return {!Array<string>} The strings with renamings applied.
+ */
+export function doRenamings(database, currVersion, newVersion, strings) {
```
I separated this part of the algorithm out into an exported function to make the system easier to test.

cpcallen replied:
Yes, this looks good, at the very least as a place to start.

cpcallen committed:
feat: Add test for getDatabase

And modify getDatabase to use the currently-in-review updated
renamings.json5 file.

BeksOmega committed:
fix: add "integration test", and some edge case tests

cpcallen committed:
feat: Initial implementation of calculateRenamings

Does not yet support getMethod / setMethod.  Probably mishandles
cases where a property becomes an export when an export of the
same name is renamed at the same time - correctness will
(incorrectly) depend on the order of the entries in database.

BeksOmega committed:
fix: add some edge cases for ordering

BeksOmega committed:
fix: use actual database of renames

cpcallen committed:
fix: More work on calculateRenamings

This causes it to pass a few more tests: 19 -> 24.

cpcallen committed:
refactor: Introduce Renamer class

The intention is to encapsulate the digested renamings so that
they can be applied to multiple files without needing to load
all those files in to memory simultaneously.

Retain doRenamings function temporarily so as not to break
the tests.

cpcallen committed:
refactor: Use Renamer directly in tests

cpcallen committed:
fix(tests): Fix test accidentally broken by previous refactor

BeksOmega committed:
feat: add basic running of commands

BeksOmega committed:
feat: add subcommand for renamings

BeksOmega committed:
fix: add test for develop version

BeksOmega committed:
fix: add tests for properly assuming versions

cpcallen committed:
refactor: Extract per-version renaming into VersionRenamer

This fixes the tests that rename things again in a later version.

cpcallen committed:
fix: Add .newExport to module path (if present)

BeksOmega committed:
fix: restoring spy on console log

BeksOmega committed:
feat: add help info for individual versions

cpcallen committed:
feat: Parse text for dotted identifiers; rename those

cpcallen committed:
fix(tests): Fix erroneous tests

Fix test failures caused by erroneous test input/expected values.
Most remaining test failures are probably actual bugs in the
implementation.

cpcallen committed:
fix: Actually apply no-change renamings

Renamings that do not result in a change of name should still be
added to .renamings_, as they will thereby prevent later,
less-specific renamings being applied to paths that should remain
unchanged.

BeksOmega committed:
fix: update version assumption tests

BeksOmega committed:
fix: remove console spy

BeksOmega committed:
fix: version coercion

BeksOmega committed:
fix: add support for develop version

BeksOmega committed:
feat: add in-place command line option

cpcallen committed:
feat: Support getMethod and setMethod

A little hacky, but works.

BeksOmega commented on plugins/migration/bin/rename.js:
```diff
@@ -246,6 +256,18 @@ class VersionRenamer {
+                'setting it.');
+          }
+          console.log('You will need to manually verify this update.');
+          return (entry.get ?? entry.set) + '()' + str.slice(entry.old.length);
```
When I was adding the `-i` option I learned that `console.log` outputs to stdout. I'm not sure if that is desirable in this case. Eg people could be doing `npx @blockly/migrate rename 6 7 my.js > new.js` and they will get these logs in their new.js file.

cpcallen replied:
I agree that that is not desirable.  We should use `process.stderr.write` or the like instead.

cpcallen commented on plugins/migration/bin/command.js:
```diff
@@ -0,0 +1,111 @@
+import semver from 'semver';
+import {Command} from 'commander';
+
+// TODO: Not sure the best way to access the JSON from node.
```
I think the code is fine, but maybe a comment like this would be useful:

```suggestion
// require() is not available in ES modules by default, so we must obtain it.
```

cpcallen commented on plugins/migration/bin/command.js:
```diff
@@ -0,0 +1,111 @@
+        'output the version of this script')
+    .usage(`${chalk.blue('<migration>')} ` +
+        `${chalk.green('<from-version> <to-version>')} [options] ` +
+        `${chalk.green('<file...>')}`);
```
Out of interest, did you pick these colours based on some standard?

BeksOmega replied:
Nope!

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+    .option(
+        '-i [suffix]',
+        'do renamings in-place, optionally create backup files with the ' +
+        'given suffix')
```
Perhaps mention that otherwise modified code will be written to stdout.

cpcallen commented on plugins/migration/bin/command.js:
```diff
@@ -0,0 +1,111 @@
+  const subCommand = new Command(name)
+      .description(description)
+      .argument('<from-version>', 'Blockly version to migrate from')
+      .argument('<to-version>', 'Blockly version to migrate to')
```
I'd suggest that this should be optional, and default to the most recently released version.  I'd be ok with requiring `--from` and `--to` or the like

BeksOmega replied:
`--from <from-version>` is now a required option rather than an argument. `--to <to-version>` is an optional argument with a required value. It defaults to `'latest'`. Working with `'latest'` and `'develop'` versions is now handled in a separate `versions.js` file so that it can be reused by future migrations.

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+      const toVersion = this.processedArgs[1];
+      const fileGlobs = this.processedArgs[2];
+      const fileNames = fileGlobs.flatMap((fileGlob) =>
+        glob.sync(fileGlob, {nodir: true, nonull: false}));
```
Why are we doing globbing?  Is this for Windows users whose shell doesn't do it for them?  I'd prefer not to glob on platforms where the shell does—it makes it difficult to migrate files that have asterisks in their name…  ;-)

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+
+      const renamer = new Renamer(await getDatabase(), fromVersion, toVersion);
+      fileNames.forEach((name) => {
+        readFile(name, 'utf8', (err, contents) => {
```
Consider using `readFileSync`, and later `writeFileSync`, since we don't actually need async here.

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+createAndAddSubCommand(
+    'rename', '>=5', 'Perform renamings for breaking changes')
+    .option(
+        '-i [suffix]',
```
Can we also accept `--in-place`?

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+ */
+function throwError(err) {
+  if (err) throw err;
+}
```
Do the errors thrown by this function contain an informative stack trace?  Do they cause the program to terminate?

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+          if (i) {
+            if (typeof i == 'string') writeFile(name + i, contents, throwError);
+            writeFile(name, newContents, throwError);
+            console.log(`Migrated renamings in ${name}`);
```
This will get printed too soon, before the write completes successfully.

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+            writeFile(name, newContents, throwError);
+            console.log(`Migrated renamings in ${name}`);
+          } else {
+            process.stdout.write(newContents);
```
Is it guaranteed that this can't result in writes becoming interleaved, even though we are not waiting for the first `write` to complete?

BeksOmega replied:
Should no longer be a problem if we write synchronously correct?

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+ * @return {!Promise<Object>} The database of renames as an object.
+ */
+export async function getDatabase() {
+  const response = await fetch(DATABASE_URL);
```
Oh yeah: this should probably be a command line option (but with the existing value as default, obviously).

BeksOmega replied:
Why?

maribethb replied:
+1, one reason is so that we can test upcoming changes to the renamings file before publishing them to this default location.

maribethb replied:
I also still think we need this.

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+ */
+const DEV_VERSION = 'develop';
+
+createAndAddSubCommand(
```
Is there a way that we could make this module side-effect free, and instead export the subcommand?  (That would also give `migrate.js` a bit more purpose.)

BeksOmega replied:
I was trying to make sure that `ROOT_COMMAND` could only have subcommands with the given form, by making it so that the only external access to it was via `createAndAddSubcommand` and `parseAndRunMigrations`. IE no `addSubCommand` function was available.

I could also create a subclass of `Command` and use type annotations to "ensure" those are the only things that get added to the `ROOT_COMMAND`... but that seems like overkill. I'll just add an `addSubCommand` function to remove side effects, and hope nobody decides to break the pattern in the future.

cpcallen commented on plugins/migration/bin/command.js:
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env node
```
Unless I'm mistaken, migrate.js is the only file in this package that's actually an executable script in its own right (i.e., should be +x and have a `#!` line); the rest are just regular JS files.  (Here and most files below.)

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+   * @param {string} version  The version to coerce.
+   * @return  {string} The coerced version.
+   */
+  static coerce(version) {
```
Perhaps `coerceVersion`?

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+   * @param {string} v2 The second version to compare.
+   * @return {number} A number indicating the relationship between the versions.
+   */
+  static compare(v1, v2) {
```
Perhaps `compareVersions`?

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+   * @return {boolean} True if the first version is less than or equal to the
+   *     second one.
+   */
+  static lte(v1, v2) {
```
I'd prefer a more descriptive name (and for `gt`), but I can't actually think of anything I like better.  :-/

BeksOmega replied:
Made these descriptive by moving them to a new versions file. The renamer now uses `versionUtils.lte` which I think is clear enough.

cpcallen commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+   * @param {string} str The string to apply the renamings in.
+   * @return {string} The file with renamings applied.
+   */
+  rename(str) {
```
I'd suggest keeping this together with `calculateRenamings`, to make it easier to read both at the same time.

cpcallen reviewed (commented):
I've had a look through the code that you wrote.  Overall looks great, though I have some suggestions.

* `migrate.js` is so small that I'd probably try to fold `command.js` into it.
* It would be good to have an automated test that actually exercises the script via the command line, perhaps via [`child_process.execSync`](https://nodejs.org/api/child_process.html#child_processexecsynccommand-options).
* `GETSTARTED.md` and `README.md` need to be updated appropriately.
* Do we want to include `manual-test-data.js`?  If so, an explanatory comment at the top about how to use it would be helpful.

Can I leave you to finish this up, and then get one of our colleagues to review?

BeksOmega commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,286 @@
+    // See the sample entry in renamings.json5 for explanation of the
+    // meaning of the different properties on database entries.
+    for (const module of entry) {
+      const oldModulePath = module.oldPath ?? module.oldName;
```
Apparently null coalescing [doesn't exist in node twelve](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator#browser_compatibility) :/ So I'm going to have to shuffle this around a bit.

BeksOmega committed:
fix: move console logs to stderr.write

BeksOmega committed:
fix: little comments

BeksOmega committed:
fix: use synchronous reads/writes

BeksOmega committed:
fix: remove globbing

BeksOmega committed:
feat: make to-version optional and have it default to latest

BeksOmega committed:
fix: remove side effects

BeksOmega committed:
fix: remove null coalescing

BeksOmega committed:
fix: update docs

BeksOmega committed:
fix: update plugin readme

BeksOmega committed:
fix: add test for renamers

maribethb commented on plugins/migration/bin/command.js:
```diff
@@ -0,0 +1,135 @@
+import semver from 'semver';
+import {Command} from 'commander';
+
+// require() is not available in ES modules by default, so we must obtain it.
```
I'm a little confused by this part -- why do we need to mix `import` and `require`? The comment should explain that.

 `@blockly/create-package` is executable via `npx` and just uses `require` directly without doing this, so I guess that's not the reason. perhaps because it only uses `require` and not `import` so it is automatically understood as a CommonJs module instead of an ES module? which leads me back to why are we mixing them?

BeksOmega replied:
> perhaps because it only uses require and not import so it is automatically understood as a CommonJs module instead of an ES module?

We instruct npm to interpret it as an ES module by adding `"type": "module"` to the `package.json`. 

> which leads me back to why are we mixing them?

We don't need to mix them, we could just use CommonJS modules, but it seems like we're trying to move toward ES modules, so we may not want to. However, `import` doesn't allow you to import JSON, which we need to do to get access to the script name and version. So we need to import require to do that.

If you want to switch everything over to CommonJS we can do that. But if we want to keep it as ES Modules afaict this is the standard solution.

maribethb commented on plugins/migration/bin/command.js:
```diff
@@ -0,0 +1,135 @@
+      if (this.args.length) {
+        showVersionHelp(this.args[0]);
+      } else {
+        ROOT_COMMAND.help();
```
What does this call? it looks like it would be an infinite loop but I don't know that much about Commander

BeksOmega replied:
It calls commanders automatically generated help. It has information about sub commands and their descriptions, but they're just organized alphabetically.

maribethb commented on plugins/migration/bin/command.js:
```diff
@@ -0,0 +1,135 @@
+ * @param {string} version The version to show migrations for.
+ */
+function showVersionHelp(version) {
+  const targetRange = `${version} - ${version}`;
```
In terms of semver ranges, I don't see how this is different than just `version`?

BeksOmega replied:
Ahah! You are correct :D I didn't think the package would accept that, but it does. Thank you Maribeth :D

maribethb commented on plugins/migration/bin/command.js:
```diff
@@ -0,0 +1,135 @@
+/**
+ * Extracts the from-version, to-version, and file names from the options and
+ * arguments.
+ * @param {Command} command The command to extract the info from.
```
`!Command`

maribethb commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,212 @@
+import JSON5 from 'json5';
+
+
+const DATABASE_URL = `https://raw.githubusercontent.com/google/blockly/develop/scripts/migration/renamings.json5`;
```
are we sure this should be develop and not master? on the one hand i guess if changes are gated by the release you're updating to, it's fine to have unreleased changes here, but on the other hand develop is a lot more active and what if we break the renamings file during the quarter? then we've broken this script "in production" despite not doing a release

i also haven't read far enough to know how the default value of "to" version is chosen to know if that might be a problem from using develop branch

BeksOmega replied:
> it's fine to have unreleased changes here, but on the other hand develop is a lot more active and what if we break the renamings file during the quarter?

Christopher wanted it to be develop so that if we forgot to add renamings to the database before a release we could add them to the database without having to make a commit on master. Wrt breaking the file, the schema that we added to core should help with that (theoretically?) But I agree, that is pretty worrisome. 

> i also haven't read far enough to know how the default value of "to" version is chosen to know if that might be a problem from using develop branch

The default value of "to" is a special value "latest" which performs all renames up to the latest released version (so up to the version before "develop" in the database

maribethb replied:
I think it should be master. Changing this should be considered the same as a "release" so if we forget to add renamings then we should accept the consequences and cherry-pick the additions into master. We don't even have to actually release a patch version if we do that, so there's not a lot of downside imo.

maribethb commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,212 @@
+        if (i) {
+          if (typeof i == 'string') writeFileSync(name + i, contents);
+          writeFileSync(name, newContents);
+          process.stderr.write(`Migrated renamings in ${name}`);
```
Why does this one go to stderr?

BeksOmega replied:
All the informative messages in `rename.js` go to stderr so that people get the messages even when they're redirecting the command output to new files. Eg `npx @blockly/migrate rename --from 7 ./file.js > ./newFile.js`

maribethb commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,212 @@
+        const newContents = renamer.rename(contents);
+        const i = this.opts().i;
+        if (i) {
+          if (typeof i == 'string') writeFileSync(name + i, contents);
```
I know we discussed allowing single-line if statements, but I thought the consensus was only for things like `if (a) return;` and not for actual statements that do things. I would strongly prefer this be on its own line, though we can revisit this as a team if opinions are split.

maribethb commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,212 @@
+
+/**
+ * Gets the database of renames.
+ * @return {!Promise<Object>} The database of renames as an object.
```
`!Promise<!Object>`

unless parse can return null? in which case `?`

i feel like we would want some error messaging here in case the file can't be parsed, instead of just crashing

maribethb commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,212 @@
+ * @param {string} newVersion The version to migrate to.
+ * @param {!Array<string>} strings The strings to apply the renamings in.
+ * @return {!Array<string>} The strings with renamings applied.
+ * @deprecated Use Renamer instance instead.
```
Why add a function that is already marked deprecated? i'm a little confused here

BeksOmega replied:
Whoops, this was how we were structuring the API before, and it should be removed how that we added the classes. Thanks for the catch!

maribethb commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,212 @@
+   * Applies the given renamings directly to a JavaScript string
+   * (presumably the contents of a developer's file).
+   * @param {string} str The string to apply the renamings in.
+   * @return {string} The file with renamings applied.
```
nit: to be consistent with the other language here, you're returning a string, not a file

maribethb commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,212 @@
+ * An object which can apply a given set of renamings for a single
+ * Blockly version to the individual dotted-identifier paths.
+ */
+class VersionRenamer {
```
This marks the end of what I've reviewed so far, I'm taking a break for now but wanted to get the above feedback in so we can work simultaneously. if anything i've said would just make more sense if i waited to review the rest of it, my bad!

maribethb commented on plugins/migration/bin/migrate.js:
```diff
@@ -0,0 +1,17 @@
+ */
+
+/**
+ * @fileoverview A script for helping to migrate between versions of Blockly.
```
I don't know what the purpose of this file is... if we create the subcommand in `rename.js` why can't we add it there? why do we call `parseAndRunMigrations` with no args?

i'm sure there are good reasons :) but could you add them to the comments in this file?

BeksOmega replied:
> why can't we add it there?

There seems to be disagreement between you and Christopher about whether modules should have side effects or not haha. That's what we did before, but Christopher wanted `rename.js` side-effect-less (which I think is reasonable) so I changed it.

> why do we call parseAndRunMigrations with no args?

It defaults to using `process.argv` if you pass no args. I'll add comments to indicate that!

BeksOmega committed:
fix: pr comments

BeksOmega committed:
fix: handling of directories, and informative output

rachel-fenichel commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,208 @@
+ * RegExp matching a dotted identifier path like "foo.bar.baz".  Note
+ * that this only matches such paths containing at least one dot, as
+ * we expect to be looking for string like "Blockly.<something>" and
+ * don't want to try to rename every singe variable and every word
```
nit: singe -> single

rachel-fenichel commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,208 @@
+      if (module.exports) {
+        for (const [oldExportName, info] of Object.entries(module.exports)) {
+          const oldExportPath =
+              info.oldPath || `${oldModulePath}.${oldExportName}`;
```
Why are you mixing template strings and string concatenation? (here vs next line and line 145).

rachel-fenichel commented on plugins/migration/bin/rename.js:
```diff
@@ -0,0 +1,208 @@
+        }
+      }
+
+      this.renamings_.push({old: oldModulePath, new: newModulePath});
```
I expected this to be directly after the lines that construct `oldModulePath` and `newModulePath`. is it down here because the order of the renamings matters?

Also, why is it safe to use `new` as a key in your object here? It's normally a reserved keyword.

BeksOmega replied:
I believe it is down here because we need to rename properties before renaming modules, otherwise we won't be able to properly detect properties that need to be renamed.

Eg:
moduleA -> moduleB
moduleA.exportA -> moduleC.exportC

If we rename moduleA first we won't be able to find moduleA.exportA because it will have become moduleB.exportA

@cpcallen  to check. If I'm right, I'll add a comment.

> Also, why is it safe to use new as a key in your object here? It's normally a reserved keyword.

I don't think there's any conflict because we're using it as a property, not a local identifier. Different grammar rules afaik.

rachel-fenichel reviewed (approved):
A few nits, but overall LGTM.

BeksOmega committed:
fix: typos, and use template strings

BeksOmega committed:
fix: move default database to master and add --database option

BeksOmega committed:
fix: add comment about rename order
